Use stricter, RFC 2141 compliant regex

mudge opened the pull request:
GitHub: https://github.com/altmetric/urn/issues/2

Be stricter with URN validation: specifically matching URNs compliant with RFC 2141:
- Starting with "urn:"
- Containing a namespace identifier that is not "urn", starting with an alphanumeric character followed by up to 31 alphanumeric or hyphen characters
- Containing a namespace specific string of permitted characters and -escaping

Note this will reject any URNs that are not %-escaped.

This does not alter the API or use the regular expression to extract the namespace identifier and namespace specific string.

mudge committed:
Use stricter, RFC 2141 compliant regex

GitHub: https://github.com/altmetric/urn/issues/2

Be stricter with URN validation: specifically matching URNs compliant
with RFC 2141:

* Starting with "urn:"
* Containing a namespace identifier that is not "urn", starting with an
  alphanumeric character followed by up to 31 alphanumeric or hyphen
  characters
* Containing a namespace specific string of permitted characters and
  %-escaping

Note this will reject any URNs that are not %-escaped.

This does not alter the API or use the regular expression to extract the
namespace identifier and namespace specific string.

mudge committed:
Update README with RFC compliance

jbilbo commented on spec/urn_spec.rb:
```diff
@@ -17,6 +18,24 @@
+    end
+
+    it 'returns false if the namespace specific string has reserved characters' do
+      expect(described_class.new('urn:foo:café')).not_to be_valid
```
:coffee:

jbilbo commented on lib/urn.rb:
```diff
@@ -1,5 +1,6 @@
 class URN
-  PATTERN = 'urn:(?!urn:)[a-z0-9\-]{1,31}:[\S]+'.freeze
+  PATTERN = %(urn:(?!urn:)[A-Za-z0-9][A-Za-z0-9\-]{1,31}:) +
```
Because that's the generic `PATTERN` (without any case-insensitive flag), should we be stricter here and change the `urn` to `[Uu][Rr][Nn]`?

mudge replied:
How about using [sub-expression options](http://ruby-doc.org/core-2.1.1/Regexp.html#class-Regexp-label-Options) to make it case insensitive?

``` ruby
  PATTERN = %{(?i:urn:(?!urn:)[a-z0-9][a-z0-9\-]{1,31}:} +
            %{(?:[a-z0-9()+,-.:=@;$_!*']|%[0-9a-f]{2})+)}.freeze
```

jbilbo replied:
Brilliant! we even fixed another issue I just noticed, we were allowing upper-case `urn` string in the NID too (e.g. `urn:URN:123`)

jbilbo replied:
Should we take out the `i` option in the `REGEX` then?

mudge replied:
Yup: I'll update it now (and add a test to forbid uppercase `urn` as a namespace identifier).

mudge committed:
Ensure the URN pattern is case-insensitive

This makes the `i` option in `REGEX` redundant so remove it and ensure
that it anchors to the start and end of the string (and not a line).

jbilbo commented:
Good work! :sparkles: 

[![LGTM](http://lgtm.in/p/yzfG0LXtn)](http://lgtm.in/i/yzfG0LXtn)
